032-projects/proposal.md

```diff
@@ -0,0 +1,274 @@
+# Summary
+
+A Concourse project is a namespace for resources, tasks, and pipelines. Projects are bootstrapped from a single resource definition and a build plan which executes when new versions of the resource are found. Projects bridge the gap between small-scale and large-scale CI workflows, helping to smoothen Concourse's learning curve.
+
+# Motivation
+
+* Smoothen the learning curve. Right now the jump from `fly execute` to `fly set-pipeline` is pretty steep. Projects can be a user's first introduction to build plans, which transitions nicely into jobs and pipelines.
+* Enable small-scale projects to have a simple Travis/Circle CI single-build workflow, without needing pipelines at all. This is a scenario where Concourse pipelines can be seen as overkill.
+* Enable medium-to-large-scale projects to fully automate their CI stack by using the project's build plan to configure pipelines via the [`set_pipeline` step](https://github.com/concourse/rfcs/pull/31).
+* Enable cross-pipeline `passed` constraints between pipelines within the project. This will allow large pipelines to be broken up into reusable templates while still supporting fan-in downstream.
+* Ensure tasks and pipelines are always in sync by having both of them come from the project instead of passing task configs as a resource throughout the pipeline.
+
+# Proposal
+
+A Concourse project allows you to bootstrap and automate your CI/CD stack by configuring a single resource which contains your project's config:
+
+```sh
+fly set-project \
+  --project booklit \
+  --type git \
+  --source uri=https://github.com/vito/booklit \
+  --source branch=projects-mock \
+  --path ci
```

Comment:
"Project version" only refers to the versions returned by the project resource's `check` action, i.e. `ref: abcdef` if you were to use a `git` resource.

I hadn't thought about having Concourse monitor the files under the given directory for change; the `--path` is just used so that Concourse knows where to load the project config from. It will rely on the `check` script and run whenever there's a new version returned by it, regardless of whether the specified path actually changed.

If there's concern over executing the `plan:` unnecessarily, it would make sense to configure the `git` resource with `paths: [ci]` in this case:

```yaml
--source paths=["ci"]
```

Alternatively the `ci` path could be split out into a separate repo like we've done (https://github.com/concourse/ci). In that case you wouldn't specify `--path` at all.

Does that make sense? :slightly_smiling_face:

Reply:
It does make sense, thanks for helping me understand. To be honest I was originally trying to understand exactly where a `project`'s `version` came from at all, but now it's conceptually clear to me 👍 thanks